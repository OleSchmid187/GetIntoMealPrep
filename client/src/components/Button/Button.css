.ui-button {
  border: none;
  border-radius: 8px;
  font-weight: bold;
  cursor: pointer;
  transition: all 0.2s ease-in-out;
}

/* Sizes */
.ui-button--small {
  padding: 0.4rem 0.8rem;
  font-size: 0.85rem;
}

.ui-button--medium {
  padding: 0.6rem 1.2rem;
  font-size: 1rem;
}

.ui-button--large {
  padding: 0.8rem 1.6rem;
  font-size: 1.1rem;
}

/* Colors (example) */
.ui-button--primary {
  background-color: var(--color-accent);
  color: var(--color-button-text);
}

.ui-button--secondary {
  background-color: #0b5e87;
  color: #ffffff;
}
.ui-button--secondary:hover {
  background-color: #0d3c6a;
}

/* Hover */
.ui-button:hover {
  transform: scale(1.05);
  opacity: 0.9;
}
