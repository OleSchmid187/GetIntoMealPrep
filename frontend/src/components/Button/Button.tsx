import React from "react";
import "./Button.css";

type ButtonProps = {
  children: React.ReactNode;
  size?: "small" | "medium" | "large";
  color?: "primary" | "secondary";
  className?: string;
  onClick?: () => void;
  type?: "button" | "submit" | "reset";
};

const Button: React.FC<ButtonProps> = ({
  children,
  size = "medium",
  color = "primary",
  className = "",
  onClick,
  type = "button",
}) => {
  const combinedClassName = [
    "ui-button",
    `ui-button--${size}`,
    `ui-button--${color}`,
    className,
  ]
    .filter(Boolean)
    .join(" ");

  return (
    <button type={type} onClick={onClick} className={combinedClassName}>
      {children}
    </button>
  );
};

export default Button;
