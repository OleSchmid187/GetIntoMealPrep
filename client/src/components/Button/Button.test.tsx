import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';
import '@testing-library/jest-dom';
import Button from './Button';

describe('Button Component', () => {
  describe('Rendering', () => {
    it('renders children correctly', () => {
      render(<Button>Click me</Button>);
      expect(screen.getByRole('button', { name: 'Click me' })).toBeInTheDocument();
    });

    it('renders with complex children', () => {
      render(
        <Button>
          <span>Icon</span> Click me
        </Button>
      );
      expect(screen.getByText('Icon')).toBeInTheDocument();
      expect(screen.getByText('Click me')).toBeInTheDocument();
    });
  });

  describe('Size Variants', () => {
    it('applies small size class', () => {
      render(<Button size="small">Small Button</Button>);
      const button = screen.getByRole('button');
      expect(button).toHaveClass('ui-button--small');
    });

    it('applies medium size class by default', () => {
      render(<Button>Default Button</Button>);
      const button = screen.getByRole('button');
      expect(button).toHaveClass('ui-button--medium');
    });

    it('applies large size class', () => {
      render(<Button size="large">Large Button</Button>);
      const button = screen.getByRole('button');
      expect(button).toHaveClass('ui-button--large');
    });
  });

  describe('Color Variants', () => {
    it('applies primary color class by default', () => {
      render(<Button>Primary Button</Button>);
      const button = screen.getByRole('button');
      expect(button).toHaveClass('ui-button--primary');
    });

    it('applies secondary color class', () => {
      render(<Button color="secondary">Secondary Button</Button>);
      const button = screen.getByRole('button');
      expect(button).toHaveClass('ui-button--secondary');
    });
  });

  describe('Type Attribute', () => {
    it('defaults to button type', () => {
      render(<Button>Button</Button>);
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute('type', 'button');
    });

    it('applies submit type', () => {
      render(<Button type="submit">Submit</Button>);
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute('type', 'submit');
    });

    it('applies reset type', () => {
      render(<Button type="reset">Reset</Button>);
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute('type', 'reset');
    });
  });

  describe('CSS Classes', () => {
    it('applies base ui-button class', () => {
      render(<Button>Button</Button>);
      const button = screen.getByRole('button');
      expect(button).toHaveClass('ui-button');
    });

    it('applies custom className', () => {
      render(<Button className="custom-class">Button</Button>);
      const button = screen.getByRole('button');
      expect(button).toHaveClass('custom-class');
    });

    it('combines all classes correctly', () => {
      render(
        <Button 
          size="large" 
          color="secondary" 
          className="custom-class extra-class"
        >
          Button
        </Button>
      );
      const button = screen.getByRole('button');
      expect(button).toHaveClass(
        'ui-button',
        'ui-button--large',
        'ui-button--secondary',
        'custom-class',
        'extra-class'
      );
    });

    it('handles empty className gracefully', () => {
      render(<Button className="">Button</Button>);
      const button = screen.getByRole('button');
      expect(button).toHaveClass('ui-button', 'ui-button--medium', 'ui-button--primary');
      expect(button.className).not.toContain('  '); // No double spaces
    });
  });

  describe('Event Handling', () => {
    it('calls onClick when clicked', async () => {
      const handleClick = vi.fn();
      const user = userEvent.setup();
      
      render(<Button onClick={handleClick}>Click me</Button>);
      const button = screen.getByRole('button');
      
      await user.click(button);
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('does not call onClick when not provided', async () => {
      const user = userEvent.setup();
      
      render(<Button>Click me</Button>);
      const button = screen.getByRole('button');
      
      // Should not throw error
      await user.click(button);
      expect(button).toBeInTheDocument();
    });

    it('handles multiple clicks', async () => {
      const handleClick = vi.fn();
      const user = userEvent.setup();
      
      render(<Button onClick={handleClick}>Click me</Button>);
      const button = screen.getByRole('button');
      
      await user.click(button);
      await user.click(button);
      await user.click(button);
      
      expect(handleClick).toHaveBeenCalledTimes(3);
    });

    it('handles keyboard navigation', async () => {
      const handleClick = vi.fn();
      const user = userEvent.setup();
      
      render(<Button onClick={handleClick}>Click me</Button>);
      const button = screen.getByRole('button');
      
      button.focus();
      await user.keyboard('{Enter}');
      expect(handleClick).toHaveBeenCalledTimes(1);
      
      await user.keyboard(' ');
      expect(handleClick).toHaveBeenCalledTimes(2);
    });
  });

  describe('Accessibility', () => {
    it('is focusable by default', () => {
      render(<Button>Focusable Button</Button>);
      const button = screen.getByRole('button');
      
      button.focus();
      expect(button).toHaveFocus();
    });

    it('has correct role', () => {
      render(<Button>Button</Button>);
      const button = screen.getByRole('button');
      expect(button).toBeInTheDocument();
    });

    it('supports aria-label through children', () => {
      render(<Button>Save Document</Button>);
      expect(screen.getByRole('button', { name: 'Save Document' })).toBeInTheDocument();
    });
  });

  describe('Form Integration', () => {
    it('submits form when type is submit', () => {
      const handleSubmit = vi.fn((e) => e.preventDefault());
      
      render(
        <form onSubmit={handleSubmit}>
          <Button type="submit">Submit Form</Button>
        </form>
      );
      
      const button = screen.getByRole('button');
      fireEvent.click(button);
      
      expect(handleSubmit).toHaveBeenCalledTimes(1);
    });

    it('resets form when type is reset', () => {
      render(
        <form>
          <input defaultValue="test" />
          <Button type="reset">Reset Form</Button>
        </form>
      );
      
      const input = screen.getByRole('textbox') as HTMLInputElement;
      const button = screen.getByRole('button');
      
      // Change input value
      fireEvent.change(input, { target: { value: 'changed' } });
      expect(input.value).toBe('changed');
      
      // Reset form
      fireEvent.click(button);
      expect(input.value).toBe('test');
    });
  });

  describe('Edge Cases', () => {
    it('handles undefined onClick gracefully', () => {
      render(<Button onClick={undefined}>Button</Button>);
      const button = screen.getByRole('button');
      
      expect(() => fireEvent.click(button)).not.toThrow();
    });

    it('handles null children', () => {
      render(<Button>{null}</Button>);
      const button = screen.getByRole('button');
      expect(button).toBeInTheDocument();
      expect(button.textContent).toBe('');
    });

    it('handles boolean children', () => {
      render(<Button>{false}</Button>);
      const button = screen.getByRole('button');
      expect(button).toBeInTheDocument();
      expect(button.textContent).toBe('');
    });

    it('handles number children', () => {
      render(<Button>{42}</Button>);
      const button = screen.getByRole('button');
      expect(button).toBeInTheDocument();
      expect(button.textContent).toBe('42');
    });
  });

  describe('Performance', () => {
    it('does not re-render unnecessarily', () => {
      const { rerender } = render(<Button>Button</Button>);
      const button = screen.getByRole('button');
      const initialClassName = button.className;
      
      // Re-render with same props
      rerender(<Button>Button</Button>);
      expect(button.className).toBe(initialClassName);
    });
  });

  describe('Snapshot Testing', () => {
    it('matches snapshot with default props', () => {
      const { container } = render(<Button>Default Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot with all props', () => {
      const { container } = render(
        <Button
          size="large"
          color="secondary"
          type="submit"
          className="custom-class"
          onClick={() => {}}
        >
          Full Props Button
        </Button>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
