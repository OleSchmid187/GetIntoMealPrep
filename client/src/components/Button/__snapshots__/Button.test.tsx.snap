// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > Snapshot Testing > matches snapshot with all props 1`] = `
<button
  class="ui-button ui-button--large ui-button--secondary custom-class"
  type="submit"
>
  Full Props Button
</button>
`;

exports[`Button Component > Snapshot Testing > matches snapshot with default props 1`] = `
<button
  class="ui-button ui-button--medium ui-button--primary"
  type="button"
>
  Default Button
</button>
`;
